actionpack/lib/action_dispatch/middleware/cookies.rb

```diff
@@ -548,63 +561,92 @@ def serializer
         def digest
           request.cookies_digest || "SHA1"
         end
-
-        def key_generator
-          request.key_generator
-        end
     end
 
-    class SignedCookieJar < AbstractCookieJar # :nodoc:
+    class SignedKeyRotatingCookieJar < AbstractCookieJar # :nodoc:
       include SerializedCookieJars
 
       def initialize(parent_jar)
         super
-        secret = key_generator.generate_key(request.signed_cookie_salt)
-        @verifier = ActiveSupport::MessageVerifier.new(secret, digest: digest, serializer: ActiveSupport::MessageEncryptor::NullSerializer)
+
+        secret = key_generator.generate_key(signed_cookie_salt)
+        @verifier = ActiveSupport::MessageVerifier.new(secret, digest: signed_cookie_digest, serializer: SERIALIZER)
+
+        Array.wrap(signed_cookie_rotate_out).each do |outbound|
+          @verifier.rotate(*expand_outbound_rotation(outbound))
+        end
+
+        if rotate_out_legacy_signed_cookies?
+          @verifier.rotate request.secret_token, serializer: SERIALIZER
+        end
       end
 
       private
         def parse(name, signed_message)
-          deserialize name, @verifier.verified(signed_message)
+          deserialize(name, @verifier.verify(signed_message) { |message|
+            return deserialize(name, message).tap { |deserialized| self[name] = { value: deserialized } }
```

Comment:
It is not clear why you need to use `return` here. It seems it is because this block is called more than once. So maybe the method should be clear about that.